File: Js/recursive_binary_search.js
const recursive_binary_search = (list, target) => {
    if(list.length === 0) {
        return false // breaking point
    } else {
        let middle = Math.floor(list.length/2)
        if(list[middle] === target) {
            return true
        } else if(list[middle] < target) {
            middle++  //+1 so the new array starts one step after the middle
            let newList = list.slice(middle)
            return recursive_binary_search(newList, target) //!Notice
        } else {
            const newList = list.slice(0, middle)
            return recursive_binary_search(newList, target)
        }
    }
}

const list = [1,2,3,4,5,6,7,8,9]
console.log(recursive_binary_search(list,0))

//!Notice : In JS we have to explicitly return the value 
// of the recursive call, unless the function returns 
//! undefine

File: Ruby/recursive_binary_search.rb
def recursive_binary_search(list, target)
    if list.length == 0
        false #base case
    else
        middle = (list.length/2).floor
        if list[middle] == target
            true #base case
        elsif list[middle] < target
            new_list = list[(middle =+ 1)..(list.length - 1)]
            recursive_binary_search(new_list, target)
        else
            new_list = list[0...middle]
            recursive_binary_search(new_list, target)
        end
    end
end

list = [1,2,3,5,6,7,8,9]
puts recursive_binary_search(list, 9)

# We can use ... or .. to create a subarray or the slice methode